core/account/utxodb/reserve.go

```diff
@@ -0,0 +1,365 @@
+	ErrInsufficient = errors.New("reservation found insufficient funds")
+
+	// ErrReserved indicates that a reservation could not be
+	// satisfied because some of the outputs were already reserved.
+	// When those reservations are finalized into a transaction
+	// (and no other transaction spends funds from the account),
+	// new change outputs will be created
+	// in sufficient amounts to satisfy the request.
+	ErrReserved = errors.New("reservation found outputs already reserved")
+)
+
+// UTXO describes an individual account UTXO.
+type UTXO struct {
+	bc.Outpoint
+	bc.AssetAmount
+	Script []byte
+
+	AccountID           string
+	ControlProgramIndex uint64
+}
+
+// Change represents reserved units beyond what was asked for.
+// Total reservation is for Amount+Source.Amount.
+type Change struct {
+	Source Source
+	Amount uint64
+}
+
+// Source describes the criteria to use when selecting UTXOs.
+type Source struct {
+	AssetID     bc.AssetID `json:"asset_id"`
+	AccountID   string     `json:"account_id"`
+	Amount      uint64
+	ClientToken *string `json:"client_token"`
+}
+
+// Reservation describes a reservation of a set of UTXOs belonging
+// to a particular account. Reservations are immutable.
+type Reservation struct {
+	ID          uint64
+	AccountID   string
+	UTXOs       []*UTXO
+	Change      []Change
+	Expiry      time.Time
+	ClientToken *string
+}
+
+func NewReserver(db pg.DB) *Reserver {
+	return &Reserver{
+		db:           db,
+		reservations: make(map[uint64]*Reservation),
+		accounts:     make(map[string]*accountReserver),
+	}
+}
+
+// Reserver implements a UTXO reserver that stores reservations
+// in-memory. It relies on the account_utxos table for the source of
+// truth of valid UTXOs but tracks which of those UTXOs are reserved
+// in-memory.
+//
+// To reduce latency and prevent deadlock, no two mutexes (either on
+// Reserver or accountReserver) should be held at the same time
+//
+// Reserver ensures idempotency of reservations until the reservation
+// expiration.
+type Reserver struct {
+	db                pg.DB
+	nextReservationID uint64
+	idempotency       idempotency.Group
+
+	reservationsMu sync.Mutex
+	reservations   map[uint64]*Reservation
+
+	accountsMu sync.Mutex
+	accounts   map[string]*accountReserver
+}
+
+// Reserve selects and reserves UTXOs according to the critera provided
+// in source. The resulting reservation expires at exp.
+func (re *Reserver) Reserve(ctx context.Context, source Source, exp time.Time) (*Reservation, error) {
+	if source.ClientToken == nil {
+		return re.reserve(ctx, source, exp)
+	}
+
+	untypedRes, err := re.idempotency.Once(*source.ClientToken, func() (interface{}, error) {
+		return re.reserve(ctx, source, exp)
+	})
+	return untypedRes.(*Reservation), err
+}
+
+func (re *Reserver) reserve(ctx context.Context, source Source, exp time.Time) (res *Reservation, err error) {
+	// Find the set of UTXOs that match this source.
+	utxos, err := re.findMatchingUTXOs(ctx, source)
```

Comment:
I'm a little concerned that `findMatchingUTXOs` will often produce far too many results.

Have you considered bundling up the rest of this function as a callback that can be passed down into the db row-scanning loop, so that that loop can exit early? It would mean turning some of the logic inside out, which would be a shame, because this is generally structured very well. Maybe stream matching utxos over a channel and tie it off when the request is satisfied?